scripts/specCoverageReport.ts

```diff
@@ -0,0 +1,424 @@
+import Table from 'cli-table';
+import dox from 'dox';
+import fs from 'fs';
+import { glob } from 'glob';
+
+const ABLY_FEATURES_SPEC_URL = 'https://raw.githubusercontent.com/ably/specification/main/textile/features.textile';
+const PATHS_TO_LOOK_FOR_TESTS = ['./test', './src/platform/react-hooks'];
+const PATHS_TO_IGNORE_FOR_TESTS = ['./test/common/ably-common'];
+
+enum SpecCoverageType {
+  spec = 'spec',
+  specpartial = 'specpartial',
+  nospec = 'nospec',
+  specskip = 'specskip',
+}
+
+interface SpecItem {
+  level: number;
+  key: string;
+  description: string;
+}
+
+interface TestSpecCoverage {
+  coverageType: SpecCoverageType.spec | SpecCoverageType.specpartial | SpecCoverageType.nospec;
+  /**
+   * Spec item key covered by the test, if applicable based on {@link coverageType} value.
+   */
+  specKey?: string;
+  /**
+   * An optional comment for tagged spec item. Looks like this in test docstrings:
+   *
+   * `@spec ID123 - [comment]`
+   */
+  coverageDescription?: string;
+  testCode: string;
+}
+
+class SpecItemNode implements SpecItem {
+  public level: number;
+  public key: string;
+  public description: string;
+  public parentNode: SpecItemNode | null = null;
+  public childNodes: SpecItemNode[] = [];
+  public specCoverages: TestSpecCoverage[] = [];
+  private _coveragePercentage: number | null = null;
+
+  constructor(specItem: SpecItem) {
+    this.level = specItem.level;
+    this.key = specItem.key;
+    this.description = specItem.description;
+  }
+
+  public isRoot(): boolean {
+    return this.parentNode == null;
+  }
+
+  public isLeaf(): boolean {
+    return this.childNodes.length === 0;
+  }
+
+  public getCoveragePercentage(): number {
+    // general rules for coverage:
+    // if the spec item is covered by the test with @spec tag - it contributes a weigth of 1 (100%).
+    // if the spec item is covered by the test with @specpartial tag - it contributes a weigth of 0.5 (50%).
```

Comment:
I’m not sure that this 50% rule is very intuitive; it seems a bit arbitrary and makes the data in the table seem more precise than it actually is. How about we instead have two columns indicating different coverage metrics:

- "Coverage (full)" — only `@spec` tags count towards this one
- "Coverage (partial)" — `@spec` or `@specpartial` tags count towards this one (i.e. if either of these tags exist for a spec point then the spec point is "covered" in this metric)

Reply:
Sounds reasonable. I agree that partial coverage is quite vague and doesn't necessarily mean that "50%" of the spec is covered by the partial test, so splitting them is probably better. Will change coverage calculation to use your suggestion

Reply:
Updated in https://github.com/ably/ably-js/pull/1811/commits/b1ab4b0c64fcec47cc5c798e498be17036d0ce7c